tutorial.md

```diff
@@ -0,0 +1,290 @@
+    let v = match data.take_value("v") {
+        Some(val) => Some(val.as_int().unwrap()),
+        None => None,
+    };
+    println!("v: {:?}, was deleted: {}",
+             v,
+             data.is_value_deleted("v"));
+    let vs = match data.take_value("vs") {
+        Some(val) => Some(val.into_vec().unwrap()),
+        None => None,
+    };
+    println!("vs: {:?}, was deleted: {}, deleted elements: {:?}",
+             vs,
+             data.is_value_deleted("vs"),
+             data.take_deleted_elements("vs"));
+    println!("_________");
+    Ok(())
+}
+```
+An example of the output:
+``` 
+_________
+Time UUID: 41183ab8-d07c-11ec-d982-2470743b0298, operation type: RowUpdate
+pk: 7, ck: 5
+v: None, was deleted: false
+vs: None, was deleted: false, deleted elements: [Int(6)]
+_________
+_________
+Time UUID: cc07e47a-d07c-11ec-ebc7-e1924abf3e16, operation type: RowUpdate
+pk: 1, ck: 2
+v: Some(8), was deleted: false
+vs: None, was deleted: false, deleted elements: []
+_________
+```
+
+<b>Note:</b> CDCRow's lifetime does not allow it to exist after the function terminates.
+If the user wants to save the data for later 
+(e.g. in a consumer that saves all consumed rows in a `vec`)
+they should map the rows to another data structure.
+## Saving progress
+User can periodically save progress while consuming CDC logs and restore it in case of some error.
+This way, the reading can start from last saved checkpoint instead of a timestamp given by the user.
+
+Progress is identified by checkpoints using the `Checkpoint` struct defined as follows:
+```rust
+#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Checkpoint {
+    pub timestamp: Duration,
+    pub stream_id: StreamID,
+    pub generation: GenerationTimestamp,
+}
+```
+
+Managing saving and restoring is used by implementing `CDCCheckpointSaver` trait defined as follows:
+```rust
+#[async_trait]
+pub trait CDCCheckpointSaver: Send + Sync {
+    async fn save_checkpoint(&self, checkpoint: &Checkpoint) -> anyhow::Result<()>;
+    async fn save_new_generation(&self, generation: &GenerationTimestamp) -> anyhow::Result<()>;
+    async fn load_last_generation(&self) -> anyhow::Result<Option<GenerationTimestamp>>;
+    async fn load_last_checkpoint(
+        &self,
+        stream_id: &StreamID,
+    ) -> anyhow::Result<Option<chrono::Duration>>;
+}
+```
+User can use any way they want to manage checkpoints, for example they can store it in a database or in a file. 
+There is a default implementation that uses ScyllaDB called `TableBackedCheckpointSaver`. 
+Its source code is located in file [checkpoints.rs](scylla-cdc/src/checkpoints.rs) and can serve as an inspiration how to write new checkpoint savers.
+
+Example of usage for `TableBackedCheckpointSaver`:
+```rust
+let user_checkpoint_saver = Arc::new(
+            TableBackedCheckpointSaver::new_with_default_ttl(session, "ks", "checkpoints")
+                .await
+                .unwrap(),
+        );
```

Comment:
These lines are indented too much with compared to the first one. The last two snippets have the same problem, too.